File: Task_1_1_2/src/test/java/ru/nsu/smolin/ScoringTest.java
package ru.nsu.smolin;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.List;
import org.junit.jupiter.api.Test;
import ru.nsu.smolin.model.Card;
import ru.nsu.smolin.model.Suit;
import ru.nsu.smolin.model.Value;
import ru.nsu.smolin.model.Scoring;


/**
 * Tests for Scoring.
 */
final class ScoringTest {

    private static Card createCard(Value v, Suit s) {
        return new Card(s, v);
    }

    @Test
    void checkSum1() {
        List<Card> hand = List.of(createCard(Value.EIGHT, Suit.SPADES),
                createCard(Value.TWO, Suit.SPADES));
        assertEquals(10, Scoring.total(hand));
    }

    @Test
    void checkSum2() {
        List<Card> hand = List.of(createCard(Value.TEN, Suit.SPADES),
                createCard(Value.JACK, Suit.SPADES));
        assertEquals(20, Scoring.total(hand));
    }

    @Test
    void checkSumWithOneAce() {
        List<Card> hand = List.of(createCard(Value.ACE, Suit.SPADES),
                createCard(Value.JACK, Suit.SPADES));
        assertEquals(21, Scoring.total(hand));
    }

    @Test
    void checkSumWithAceAndTotalAbove21() {
        List<Card> hand = List.of(
                createCard(Value.TEN, Suit.SPADES),
                createCard(Value.ACE, Suit.SPADES),
                createCard(Value.TWO, Suit.HEARTS)
        );
        assertEquals(13, Scoring.total(hand));
    }

    @Test
    void checkSumWithMultipleAces() {
        List<Card> hand = List.of(createCard(Value.ACE, Suit.SPADES),
                createCard(Value.ACE, Suit.HEARTS));
        assertEquals(12, Scoring.total(hand));
    }
}

File: Task_1_1_2_Kotlin/src/test/kotlin/ru/nsu/smolin/ScoringTest.kt
package ru.nsu.smolin

import org.junit.jupiter.api.Test
import kotlin.test.assertEquals

private fun c(v: Value, s: Suit) = Card(s, v)

class ScoringTest {
    @Test
    fun `check sum 1`() {
        val hand = listOf(c(Value.EIGHT, Suit.SPADES), c(Value.TWO, Suit.SPADES))
        assertEquals(10, Scoring.total(hand))
    }

    @Test
    fun `check sum 2`() {
        val hand = listOf(c(Value.TEN, Suit.SPADES), c(Value.JACK, Suit.SPADES))
        assertEquals(20, Scoring.total(hand))
    }

    @Test
    fun `check sum with one ace`() {
        val hand = listOf(c(Value.ACE, Suit.SPADES), c(Value.JACK, Suit.SPADES))
        assertEquals(21, Scoring.total(hand))
    }
    @Test
    fun `check sum with ace and total above 21`() {
        val hand = listOf(c(Value.TEN, Suit.SPADES), c(Value.ACE, Suit.SPADES), c(Value.TWO, Suit.HEARTS))
        assertEquals(13, Scoring.total(hand))
    }

    @Test
    fun `check sum with multiple aces`() {
        val hand = listOf(c(Value.ACE, Suit.SPADES), c(Value.ACE, Suit.HEARTS))
        assertEquals(12, Scoring.total(hand))
    }
}